NetworkingSampleApp/NetworkingSampleApp/Scenes/Upload/UploadItemViewModel.swift

```diff
@@ -0,0 +1,81 @@
+//
+//  UploadItemViewModel.swift
+//  NetworkingSampleApp
+//
+//  Created by Tony Ngo on 12.06.2023.
+//
+
+import Foundation
+
+@MainActor
+final class UploadItemViewModel: ObservableObject {
+    @Published private(set) var progress: Double = 0
+    @Published private(set) var formattedProgress: String = ""
+    @Published private(set) var isPaused = false
+    @Published private(set) var isCancelled = false
+    @Published private(set) var isRetryable = false
+
+    var stateTitle: String {
+        isCancelled
+            ? "Cancelled"
+            : isCompleted ? "Completed" : formattedProgress
+    }
+
+    var isCompleted: Bool { progress == 100 }
+
+    let fileName: String
+    let totalProgress = 100.0
+
+    private let item: UploadItem
+    private let uploadService: UploadService
+
+    init(item: UploadItem, uploadService: UploadService) {
+        self.item = item
+        self.fileName = item.fileName
+        self.uploadService = uploadService
+    }
+}
+
+extension UploadItemViewModel {
+    func observeProgress() async {
+        let uploadStateStream = await uploadService.uploadStateStream(for: item.id)
+        for await state in uploadStateStream {
+            progress = state.fractionCompleted * 100
+            formattedProgress = String(format: "%.2f", progress) + "%"
+            isPaused = state.isSuspended
+            isCancelled = state.cancelled
+            isRetryable = state.cancelled || state.timedOut || state.error != nil
+        }
+    }
+
+    func pause() {
+        Task {
+            await uploadService.pause(taskId: item.id)
+            isPaused = true
```

Comment:
Do you mean the `isPaused` property? This was from me just to make sure that the view model is in the correct state, otherwise the upload task's state is available in its stream here — a new state is available whenever the upload task's state changes, in this case being paused/suspended.

```swift
for await state in uploadStateStream {
    isPaused = state.isSuspended
}
```